pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,382 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloud
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"k8s.io/api/admissionregistration/v1alpha1"
+	"k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	clientretry "k8s.io/kubernetes/pkg/client/retry"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/aws"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+	kubeletapis "k8s.io/kubernetes/pkg/kubelet/apis"
+	vol "k8s.io/kubernetes/pkg/volume"
+)
+
+var UpdatePVLabelsBackoff = wait.Backoff{
+	Steps:    20,
+	Duration: 50 * time.Millisecond,
+	Jitter:   1.0,
+}
+
+const initializerName = "pvlabel.kubernetes.io"
+const initialzerConfigName = "pvlabel-initconfig"
+
+// PersistentVolumeLabelController handles adding labels to persistent volumes when they are created
+type PersistentVolumeLabelController struct {
+	ebsVolumes       aws.Volumes
+	gceCloudProvider *gce.GCECloud
+	cloud            cloudprovider.Interface
+	mutex            sync.Mutex
+	kubeClient       clientset.Interface
+	pvlController    cache.Controller
+
+	syncHandler func(vol *v1.PersistentVolume) error
+
+	// queue is where incoming work is placed to de-dup and to allow "easy" rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+}
+
+// NewPersistentVolumeLabelController creates a PersistentVolumeLabelController object
+func NewPersistentVolumeLabelController(
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface) *PersistentVolumeLabelController {
+	pvlc := &PersistentVolumeLabelController{
+		cloud:      cloud,
+		kubeClient: kubeClient,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pvLabels"),
+	}
+	pvlc.syncHandler = pvlc.AddLabels
+	_, pvlc.pvlController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.IncludeUninitialized = true
+				return kubeClient.CoreV1().PersistentVolumes().List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.IncludeUninitialized = true
+				return kubeClient.CoreV1().PersistentVolumes().Watch(options)
+			},
+		},
+		&v1.PersistentVolume{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cast := obj.(*v1.PersistentVolume)
+				pvlc.queue.Add(cast)
```

Comment:
We usually add the value from `cache.MetaNamespaceKeyFunc` as they key, and then in the sync handler, retrieve the key from the queue and use a lister to get the corresponding object out of the shared cache. This is because there might be multiple enqueues for the same key before the worker is able to pop, and you should operate on the object in question once at its current state in the cache. Right @deads2k @liggitt?